### PR TITLE
mpr121: Add GPIO support

### LINKED ISSUE #1838: Support for IO Pins on MPR121
**Describe the problem you have/What new integration you would like**

The MPR121 chip not only supports 12 channels of touch sensors: 8 of its pins can also be used as LED driver or GPIO pins. 
The MPR121 EspHome component on the other hand only supports the touch binary sensors.

**Please describe your use case for this integration and alternatives you've tried:**

It would be awesome to be able to also use the pins as e.g. binary switch outputs, because the chip supports it.
Regarding alternatives, there aren't really any feasible alternatives because of the MPR121's rich feature set.
(and because I just bought like 30 of them for a project)

**Additional context**

edit: they can even do PWM, here's a datasheet going into detail regarding the Multifunction Pins: https://www.nxp.com/docs/en/application-note/AN3894.pdf
But it seems they can only PWM at 4bit (16 steps) resolution, support for this would probably be lower priority.

### PR DESCRIPTION
**Pull request by polyfloyd**
# What does this implement/fix?

The MPR121 touch controller also supports alternate modes of operation as GPIO controller on the upper 8 of its 12 input channels.

This patch makes it possible to configure such channels as GPIO.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes [esphome/feature-requests/issues/1838](https://github.com/esphome/feature-requests/issues/1838)

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3845

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

i2c:
  sda: 26
  scl: 27

mpr121:
  id: mpr121_component
  address: 0x5A
  touch_debounce: 1
  release_debounce: 1
  touch_threshold: 10
  release_threshold: 7

binary_sensor:
  - platform: mpr121
    id: touch_LEFT
    channel: 7
    name: "Touch Left"

output:
  - id: ext_5v_power
    platform: gpio
    pin:
      mpr121_id: mpr121_component
      number: 10
      mode: OUTPUT
      inverted: true
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs). -> https://github.com/esphome/esphome-docs/pull/3845

### CODE REVIEW

**Comment by polyfloyd** [comment]
Alright! This is ready for a review :)

**Comment by jesserockz** [comment]
@polyfloyd I took it upon myself to use this platform as a testbed to do some crazy validation =)

Now you cannot use the wrong channel as a pin while there is a higher binary sensor channel being used.

I cannot test it on device, but compilation tests are working locally.

Let me know how it goes =)

Jesse

**Comment by polyfloyd** [comment]
Nope! It panics:

```[11:41:28]Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled.
[11:41:28]
[11:41:28]Core  1 register dump:
[11:41:28]PC      : 0x400d6b34  PS      : 0x00060030  A0      : 0x800d6b95  A1      : 0x3ffb25f0
WARNING Decoded 0x400d6b34: esphome::mpr121::MPR121BinarySensor::setup() at /home/polyfloyd/Projects/esphome-config/.esphome/build/geiger/src/esphome/components/mpr121/binary_sensor/mpr121_binary_sensor.cpp:7
[11:41:28]A2      : 0x3ffb34f4  A3      : 0x0000005e  A4      : 0x3ffb261c  A5      : 0x00000001
[11:41:28]A6      : 0x00000001  A7      : 0x00000001  A8      : 0x8017eb76  A9      : 0x00000000
[11:41:28]A10     : 0x00000000  A11     : 0x00000002  A12     : 0x3ffb25f0  A13     : 0x00000002
[11:41:28]A14     : 0x00000001  A15     : 0x00000001  SAR     : 0x00000004  EXCCAUSE: 0x0000001c
[11:41:28]EXCVADDR: 0x00000025  LBEG    : 0x4008976c  LEND    : 0x40089782  LCOUNT  : 0xffffffff
[11:41:28]
[11:41:28]
[11:41:28]Backtrace:0x400d6b31:0x3ffb25f00x400d6b92:0x3ffb2620 0x400d6ee5:0x3ffb2640 0x4017f629:0x3ffb2660 0x4017f6cd:0x3ffb2680 0x400e1438:0x3ffb26a0 0x400e3f79:0x3ffb26d0 0x400ea052:0x3ffb2820
WARNING Found stack trace! Trying to decode it
WARNING Decoded 0x400d6b31: unsigned char esphome::optional<unsigned char>::value_or<unsigned char>(unsigned char const&) const at /home/polyfloyd/Projects/esphome-config/.esphome/build/geiger/src/esphome/core/optional.h:93
 (inlined by) esphome::mpr121::MPR121BinarySensor::setup() at /home/polyfloyd/Projects/esphome-config/.esphome/build/geiger/src/esphome/components/mpr121/binary_sensor/mpr121_binary_sensor.cpp:7
WARNING Decoded 0x400d6b92: non-virtual thunk to esphome::mpr121::MPR121BinarySensor::setup()
WARNING Decoded 0x400d6ee5: esphome::mpr121::MPR121Component::setup() at /home/polyfloyd/Projects/esphome-config/.esphome/build/geiger/src/esphome/components/mpr121/mpr121.cpp:26 (discriminator 2)
WARNING Decoded 0x4017f629: esphome::Component::call_setup() at /home/polyfloyd/Projects/esphome-config/.esphome/build/geiger/src/esphome/core/component.cpp:78
WARNING Decoded 0x4017f6cd: esphome::Component::call() at /home/polyfloyd/Projects/esphome-config/.esphome/build/geiger/src/esphome/core/component.cpp:104
WARNING Decoded 0x400e1438: esphome::Application::setup() at /home/polyfloyd/Projects/esphome-config/.esphome/build/geiger/src/esphome/core/application.cpp:38
WARNING Decoded 0x400e3f79: setup() at /home/polyfloyd/Projects/esphome-config/.esphome/build/geiger/geiger.yaml:301
WARNING Decoded 0x400ea052: loopTask(void*) at /home/polyfloyd/.platformio/packages/framework-arduinoespressif32@3.20005.220925/cores/esp32/main.cpp:42
[11:41:28]
[11:41:28]
[11:41:28]
[11:41:28]
[11:41:28]ELF file SHA256: 0000000000000000
[11:41:28]
[11:41:28]Rebooting...
[11:41:28]ets Jun  8 2016 00:22:57
[11:41:28]
[11:41:28]rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
[11:41:28]configsip: 0, SPIWP:0xee
[11:41:28]clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
[11:41:28]mode:DIO, clock div:2
[11:41:28]load:0x3fff0030,len:1184
[11:41:28]load:0x40078000,len:13132
[11:41:28]load:0x40080400,len:3036
[11:41:28]entry 0x400805e4
```

The pin conflict validation at compile time is really nice though :)

**Comment by polyfloyd** [comment]
Also, the `mpr121_id` field of the gpio output config has changed to `mpr121`. Was this intentional? Having such references be prefixed by `_id` seems the common pattern esphome, e.g. `uart_id`

**Comment by polyfloyd** [comment]
Fixed it :P

Both touch and GPIO seems to be working well!

**Comment by jesserockz** [comment]
Ah woops, Yeah I guess I forgot that part. Although the set_parent function is already provided by the Parented class

**Comment by jesserockz** [comment]
> Also, the `mpr121_id` field of the gpio output config has changed to `mpr121`. Was this intentional? Having such references be prefixed by `_id` seems the common pattern esphome, e.g. `uart_id`

For component platforms yes, for gpio pins its just the name of the component without `_id` suffix

**Comment by polyfloyd** [comment]
Ah! Good to know

**Comment by polyfloyd** [comment]
I have updated the corresponding documentation

File: esphome/components/mpr121/__init__.py [hunk 1]
```diff
@@ -28,18 +41,87 @@
...
     .extend(cv.COMPONENT_SCHEMA)
     .extend(i2c.i2c_device_schema(0x5A))
 )
 
 
+def _final_validate(config):
+    fconf = fv.full_config.get()
+    max_touch_channel = 3
+    if CONF_BINARY_SENSOR in fconf:
+        for binary_sensor in fconf[CONF_BINARY_SENSOR]:
```
Comment (jesserockz) [inline esphome/components/mpr121/__init__.py]:
```suggestion
    if binary_sensors := fconf.get(CONF_BINARY_SENSOR) is not None:
        for binary_sensor in binary_sensors:
```

File: esphome/components/mpr121/__init__.py [hunk 2]
```diff
@@ -28,18 +41,87 @@
...
     )
     .extend(cv.COMPONENT_SCHEMA)
     .extend(i2c.i2c_device_schema(0x5A))
 )
 
 
+def _final_validate(config):
+    fconf = fv.full_config.get()
+    max_touch_channel = 3
+    if binary_sensors := fconf.get(CONF_BINARY_SENSOR) is not None:
```
Comment (jesserockz) [inline esphome/components/mpr121/__init__.py]:
```suggestion
    if (binary_sensors := fconf.get(CONF_BINARY_SENSOR)) is not None:
```

**Review by jesserockz — APPROVED** [review]

**Comment by polyfloyd** [comment]
Thanks!

**Pull request merged**